File: server/src/main/java/model/item/Weapon.java
package model.item;

import javafx.beans.binding.Bindings;
import javafx.beans.binding.IntegerBinding;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;
import manager.Database;

public class Weapon extends Equipment {

    private float initiative;

    public float getInitiativeModifier() {
        return initiativeModifier;
    }

    public void setInitiativeModifier(float initiativeModifier) {
        this.initiativeModifier = initiativeModifier;
    }

    private float initiativeModifier;
    private String dice;
    private final IntegerProperty damage;
    protected final IntegerBinding damageWithWear;
    private int hit;
    private Boolean isShield = null;

    public Weapon() {
        super();
        this.initiative = 0f;
        this.dice = "";
        this.damage = new SimpleIntegerProperty(0);
        this.damageWithWear = Bindings.createIntegerBinding(() -> Math.max(0, getDamage() - getWearStage()), damage, wearStage);
        this.hit = 0;
    }

    @Override
    protected boolean shouldBreak() {
        return getWearStage() > getDamage();
    }

    @Override
    public Weapon copy() {
        Weapon weapon = (Weapon) super.copy();
        weapon.setInitiative(this.getInitiative());
        weapon.setInitiativeModifier(this.getInitiativeModifier());
        weapon.setDice(this.getDice());
        weapon.setDamage(this.getDamage());
        weapon.setHit(this.getHit());

        return weapon;
    }

    public boolean isShield() {
        if (isShield == null) {
            isShield = Database.shieldTypes.stream().anyMatch(type -> type.equals(this.getSubtype()));
        }

        return isShield;
    }

    public float getInitiative() {
        return initiative;
    }

    public void setInitiative(float initiative) {
        this.initiative = initiative;
    }

    public String getDice() {
        return dice;
    }

    public void setDice(String dice) {
        this.dice = dice;
    }

    public int getDamage() {
        return damage.get();
    }

    public IntegerProperty damageProperty() {
        return damage;
    }

    public void setDamage(int damage) {
        this.damage.set(damage);
    }

    public IntegerBinding damageWithWearBinding() {
        return damageWithWear;
    }

    public int getHit() {
        return hit;
    }

    public void setHit(int hit) {
        this.hit = hit;
    }

    @Override
    public boolean equals(Object o) {

        if (!(o instanceof Weapon) || !super.equals(o)) {
            return false;
        }

        Weapon other = (Weapon) o;

        return this.getInitiative() == other.getInitiative() && this.getDice().equals(other.getDice())
                && this.getDamage() == other.getDamage() && this.getHit() == other.getHit() && this.getInitiativeModifier() == other.getInitiativeModifier();
    }
}

File: app/src/main/java/de/pnp/manager/app/model/Weapon.kt
package de.pnp.manager.app.model

import de.pnp.manager.model.ICurrency
import de.pnp.manager.model.IRarity
import de.pnp.manager.model.item.IEquipment
import de.pnp.manager.model.item.IWeapon
import de.pnp.manager.model.upgrade.IUpgrade
import java.util.ArrayList

class Weapon : IWeapon, Equipment() {
    private var isShield: Boolean = false
    private var initiative: Float = 0.0f
    private var dice: String = ""
    private var damage: Int = 0
    private var hit: Int = 0
    private var initiativeModifier: Float = 0.0f

    override fun copy(): IWeapon {
        val weapon = super.copy() as Weapon
        weapon.damage = this.damage
        weapon.isShield = this.isShield
        weapon.initiative = this.initiative
        weapon.dice = this.dice
        weapon.hit = this.hit
        weapon.initiativeModifier = this.initiativeModifier
        return weapon
    }

    override fun isShield(): Boolean {
        return isShield
    }

    override fun getInitiativeModifier(): Float {
        return this.initiativeModifier
    }

    override fun setInitiativeModifier(initiativeModifier: Float) {
        this.initiativeModifier = initiativeModifier
    }

    fun setIsShield(value: Boolean) {
        this.isShield = value
    }

    override fun getInitiative(): Float {
        return initiative
    }

    override fun setInitiative(initiative: Float) {
        this.initiative = initiative
    }

    override fun getDice(): String {
        return dice
    }

    override fun setDice(dice: String) {
        this.dice = dice
    }

    override fun getDamage(): Int {
        return damage
    }

    override fun setDamage(damage: Int) {
        this.damage = damage
    }

    override fun getHit(): Int {
        return hit
    }

    override fun setHit(hit: Int) {
        this.hit = hit
    }
}